flang/lib/Lower/Bridge.cpp

```diff
@@ -1621,10 +1621,8 @@ class FirConverter : public Fortran::lower::AbstractConverter {
               if (isCharacterCategory(lhsType->category())) {
                 // Fortran 2018 10.2.1.3 p10 and p11
                 // Generating value for lhs to get fir.boxchar.
-                Fortran::lower::ExpressionContext context;
-                auto lhs = fir::getBase(genExprAddr(assign.lhs));
-                auto rhs = createSomeExtendedExpression(
-                    toLocation(), *this, assign.rhs, localSymbols, context);
+                auto lhs = genExprAddr(assign.lhs);
+                auto rhs = genExprValue(assign.rhs);
```

Comment:
rhs could also probably also use  `genExprAddr` here  (given CharacterExprHelper::createAssign does not work at the value level). That would save the trouble around `genLoad`.
Your `genLoad` update for constant length character is clean, but the other cases sounds more annoying. Returning emboxchar seems like the only thing we can do, but I feel it's debatable whether or not `genLoad` should return something that is still tied to the underlying memory.

However working with Sequence instead of loops when possible might actually help later passes (Though the current assign code indeed still dumbly always allocate the rhs in a temp and uses a loop). We could actually easily save the explicit rhs temp copy in this case and leave later translation passes take care of the aliasing analysis.

Reply:
Right. I'm planning on tinkering with this more. Thanks for the ideas.